### ISSUE TITLE
How to override configs for `Optiona[Dict[xx, xx]]` in cmd?

### ISSUE THREAD

**Issue by yxlao** [brentyi/tyro#52]
### Case 1: `Dict`
The following test code works:
```python
def test_normal_dict() -> None:
    def main(x: Dict[str, int] = {"a": 5, "b": 5}):
        return x

    assert tyro.cli(main, args=[]) == {"a": 5, "b": 5}
    assert tyro.cli(main, args="--x.a 3 --x.b 7".split(" ")) == {
        "a": 3,
        "b": 7,
    }
```
For example, we can override arguments with:
```bash
--x.a 3 --x.b 7
```

### Case 2: `Optional[Dict]`

Now, if the given config is `Optional[Dict[str, int]]`, how shall we override its parameters? 
```python
def test_optional_dict() -> None:
    def main(x: Optional[Dict[str, int]] = {"a": 5, "b": 5}):
        return x

    assert tyro.cli(main, args=[]) == {"a": 5, "b": 5}
    # The following assert will fail.
    assert tyro.cli(main, args="--x.a 3 --x.b 7".split(" ")) == {
        "a": 3,
        "b": 7,
    }
```

For example, this is used in 

https://github.com/nerfstudio-project/nerfstudio/blob/081bf82bf594b6eab52410b46e6cbd510ce82eea/nerfstudio/models/base_model.py#L47-L48

**Comment by brentyi** [comment]
Hi!

The usage should match the help message:

![image](https://user-images.githubusercontent.com/6992947/236638152-0d5b1419-5369-48d3-92a9-ef4d6a22a5ba.png)

- Case 1: when we have `x: Dict[str, int] = {"a": 5, "b": 5}`, we can pass in `--x.a 3 --x.b 7`.
- Case 2: To get the same effect in `x: Optional[Dict[str, int]] = {"a": 5, "b": 5}`, we have to pass in `--x a 3 b 7`.
    - We can also pass in `--x None` to set `x=None`.

Where the script I'm running is:
```python
from typing import Dict, Optional
import tyro


def main(x: Optional[Dict[str, int]] = {"a": 5, "b": 5}):
    return x

tyro.cli(main)
```

There were unfortunately a few tradeoffs + inconsistencies I had to weigh when I was figuring out what to do with dictionary types. Hopefully the `None` case highlights why this particular decision was made: if we populated the parser with case-1 style arguments for fields annotated as `Optional[Dict[K, V]]`, there'd be no way to set the value of `x` to `None`. On the other hand, with the case 1-style arguments there's no way to add new keys to the dictionary. I have ideas for how to improve all of this, but unfortunately pretty time-limited these days so don't know if I'll be able to execute them.
